### PR TITLE
Logstash loader tool

### PR DESCRIPTION
**Pull request by timcharper**

### CODE REVIEW

File: tools/logstash/README.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+Elasticsearch. This should enable users to analyze logs with Kibana.
+
+## Generating LogStash config
+
+### DCOS Bundle
+
+LogStash config can be generated for a DCOS bundle using the following command:
+
+```
+bin/target.sc <path-to-diagnostic-bundle>
```
Comment (jeschkies) [inline tools/logstash/README.md]: I do like this approach but find the name `target` confusing. How about `bin/elk generate-config <path-to-diagnostic-bundle>`?

File: tools/logstash/README.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+## Running ElasticSearch, Kibana, and LogStash
+
+In a separate terminal EACH, run the following commands, in this order:
+
+```
+elasticsearch
+
+kibana
+
+logstash -f target/loading
```
Comment (jeschkies) [inline tools/logstash/README.md]: Is the path for the generated config files hard coded?

File: tools/logstash/README.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+Logstash will begin populating ElasticSearch with your bundles log data.
+
+Open a browser to http://localhost:5601 to use Kibana
+
+
+## Clearing data
+
+You can purge the logs from ElasticSearch using this command:
+
+`bin/clear-indices.sh`
```
Comment (jeschkies) [inline tools/logstash/README.md]: ElasticSearch has to be running, right?

File: tools/logstash/bin/target [hunk 4]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
```
Comment (jeschkies) [inline tools/logstash/bin/target]: As mentioned above, I find the name `target` confusing.

File: tools/logstash/conf/filter-marathon-1.4.x.conf [hunk 5]
```diff
@@ -0,0 +1,336 @@
+filter {
```
Comment (jeschkies) [inline tools/logstash/conf/filter-marathon-1.4.x.conf]: Did you write these?

File: tools/logstash/lib/logformat.sc [hunk 6]
```diff
@@ -0,0 +1,34 @@
+trait LogFormat {
+  def matches(s: String): Boolean
+  val codec: String
+  val unframe: String
+}
+
+object DcosLogFormat extends LogFormat {
+  val regexPrefix = "^[0-9]{4}-[0-9]{2}-[0-9]{2} [0-9]{2}:[0-9]{2}:[0-9]{2}:".r
```
Comment (jeschkies) [inline tools/logstash/lib/logformat.sc]: Could you give an example of what is matched?

File: tools/logstash/lib/main.sc [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+import scala.concurrent.{Await, Future}
+import scala.util.{Try, Success, Failure}
+
+@tailrec def await[T](f: Future[T]): T = f.value match {
+  case None =>
+    Thread.sleep(10)
+    await(f)
+  case Some(v) =>
+    v.get
+}
```
Comment (jeschkies) [inline tools/logstash/lib/main.sc]: Interesting. I've done it with

```scala
val done = Promise[Done]
done.completeWith(f)
while (!done.isCompleted) { Thread.sleep(100) }
```

probably boils down to the same thing 🙂.

File: tools/logstash/lib/main.sc [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        } yield codec).headOption
+
+        maybeCodec match {
+          case Some(codec) => codec
+          case _ =>
+            println(s"Couldn't find a codec for these lines:")
+            println()
+            linesSample.foreach(println)
+            // sys.exit(1)
+            ???
```
Comment (jeschkies) [inline tools/logstash/lib/main.sc]: What is this about?

File: tools/logstash/lib/main.sc [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+def writeFiles(entries: (Path, String)*): Unit = {
+  entries.foreach { case (path, contents) =>
+    write.over(path, contents)
+    println(s"Wrote ${path}")
+  }
+}
+
+def generateTargetBundle(path: Path): Unit = {
```
Comment (jeschkies) [inline tools/logstash/lib/main.sc]: What is path?

File: tools/logstash/README.md (hunk 2 above)
Comment (timcharper) [reply tools/logstash/README.md]: Yes

File: tools/logstash/README.md (hunk 1 above)
Comment (timcharper) [reply tools/logstash/README.md]: Okay if we keep it as is? I personally like the name, and don't find it confusing

File: tools/logstash/README.md (hunk 3 above)
Comment (timcharper) [reply tools/logstash/README.md]: ✅ Updated docs

File: tools/logstash/conf/filter-marathon-1.4.x.conf (hunk 5 above)
Comment (timcharper) [reply tools/logstash/conf/filter-marathon-1.4.x.conf]: Yes

File: tools/logstash/lib/logformat.sc (hunk 6 above)
Comment (timcharper) [reply tools/logstash/lib/logformat.sc]: ✅

File: tools/logstash/lib/main.sc (hunk 7 above)
Comment (timcharper) [reply tools/logstash/lib/main.sc]: yep :)

File: tools/logstash/lib/main.sc (hunk 8 above)
Comment (timcharper) [reply tools/logstash/lib/main.sc]: `???` throws not implemented exception. Because codec not implemented.

File: tools/logstash/lib/main.sc (hunk 9 above)
Comment (timcharper) [reply tools/logstash/lib/main.sc]: ✅

File: tools/logstash/conf/filter-marathon-1.4.x.conf (hunk 5 above)
Comment (jeschkies) [reply tools/logstash/conf/filter-marathon-1.4.x.conf]: 😱

File: tools/logstash/lib/main.sc (hunk 8 above)
Comment (jeschkies) [reply tools/logstash/lib/main.sc]: Hm, ok. I'd be more specific but let's merge this 🙂.

File: tools/elk/README.md [hunk 10]
```diff
@@ -0,0 +1,54 @@
+# Analyzing Marathon with ELK Stack
+
+Marathon comes with a basic set of grok filter for Logstash to load logs into
+Elasticsearch. This should enable users to analyze logs with Kibana.
+
+## Generating LogStash config
```
Comment (alenkacz) [inline tools/elk/README.md]: it's Logstash not LogStash - just a cosmetic... :)

Comment (alenkacz) [reply tools/elk/README.md]: but if you agree, please fix it also on other places where that appears

File: tools/elk/README.md [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+### DCOS Bundle
+
+LogStash config can be generated for a DCOS bundle using the following command:
+
+```
+bin/target bundle <path-to-diagnostic-bundle>
+```
+
+Command will attempt to detect log line format (include date, etc.), and generate the appropriate LogStash input
+directives such that the hostname field will be populated in ElasticSearch.
```
Comment (alenkacz) [inline tools/elk/README.md]: Also it's Elasticsearch not ElasticSearch

File: tools/elk/README.md (hunk 10 above)
Comment (timcharper) [reply tools/elk/README.md]: thanks for pointing this out! Fixed.

**Comment by timcharper** [comment]
I took a brief stab at running in Docker but had issues getting Elasticsearch to run in Docker for Mac. It kept crashing.

**Pull request merged**